### PR TITLE
(proof-new) Fixes and improvements for smt proof postprocessor

### PR DESCRIPTION
**Pull request by ajreynol**
This includes several subtle issues encountered in the past month based on our regressions.

It also improves the expansion forms of MACRO_ rules to use EQ_RESOLVE instead of a TRUE_INTRO/TRUE_ELIM scheme. This is both more compact and avoids cyclic proofs for some corner cases of proofs with Boolean constant equalites.

### CODE REVIEW

File: src/smt/proof_post_processor.cpp [hunk 1]
```diff
@@ -214,73 +224,87 @@ Node ProofPostprocessCallback::expandMacros(PfRule id,
...
-    // transitivity if necessary
+    if (reqWitness)
+    {
+      Node weq = addProofForWitnessForm(conc[1], cdp);
+      Trace("smt-proof-pp-debug") << "...weq is " << weq << std::endl;
+      if (addToTransChildren(weq, tchildren))
+      {
+        // toWitness(apply_SR(t)) = apply_SR(toWitness(apply_SR(t)))
+        // rewrite again, don't need substitution. Also we always use the
+        // default rewriter, due to the definition of MACRO_SR_PRED_INTRO.
```
Comment (HanielB) [inline src/smt/proof_post_processor.cpp:256]: What is it in the definition of MACRO_SR_PRED_INTRO that enforces the use of the default rewriter?

File: src/smt/proof_post_processor.cpp [hunk 2]
```diff
@@ -309,10 +334,11 @@ Node ProofPostprocessCallback::expandMacros(PfRule id,
             << "transform toWitness (" << r << "): " << weq << std::endl;
         if (addToTransChildren(weq, tchildrenr))
         {
-          sargs[0] = weq[1];
           // toWitness(apply_SR(t)) = apply_SR(toWitness(apply_SR(t)))
-          // rewrite again, don't need substitution
-          Node weqr = expandMacros(PfRule::MACRO_SR_EQ_INTRO, {}, sargs, cdp);
+          // rewrite again, don't need substitution. Also, we always use the
+          // default rewriter, due to the definition of MACRO_SR_PRED_TRANSFORM.
```
Comment (HanielB) [inline src/smt/proof_post_processor.cpp:339]: ditto

File: src/smt/proof_post_processor.cpp [hunk 3]
```diff
@@ -404,7 +433,8 @@ Node ProofPostprocessCallback::expandMacros(PfRule id,
           // add previous rewrite steps
           for (unsigned j = 0, nvars = vvec.size(); j < nvars; j++)
           {
-            tcg.addRewriteStep(vvec[j], svec[j], pgs[j]);
+            // not necessarily closed
```
Comment (HanielB) [inline src/smt/proof_post_processor.cpp:436]: please clarify

File: src/smt/proof_post_processor.cpp [hunk 4]
```diff
@@ -443,14 +473,19 @@ Node ProofPostprocessCallback::expandMacros(PfRule id,
       TConvProofGenerator tcpg(d_pnm, nullptr, TConvPolicy::ONCE);
       for (unsigned j = 0, nvars = vvec.size(); j < nvars; j++)
       {
-        tcpg.addRewriteStep(vvec[j], svec[j], pgs[j]);
+        // not necessarily closed
```
Comment (HanielB) [inline src/smt/proof_post_processor.cpp:476]: ditto

File: src/expr/proof_rule.h [hunk 5]
```diff
@@ -129,14 +129,21 @@ enum class PfRule : uint32_t
...
-  //   Rewriter{idr}(toWitness(F)*sigma{ids}(Fn)*...*sigma{ids}(F1)) == true
-  // where ids and idr are method identifiers.
+  //   Rewriter{idr}(F*sigma{ids}(Fn)*...*sigma{ids}(F1)) == true
+  // where ids and idr are method identifiers. 
   //
-  // Notice that we apply rewriting on the witness form of F, meaning that this
+  // More generally, this rule also holds when:
+  //   Rewriter::rewrite(toWitness(F')) == true
+  // where F' is the result of the left hand side of the equality above. Here,
+  // notice that we apply rewriting on the witness form of F, meaning that this
```
Comment (HanielB) [inline src/expr/proof_rule.h:138]: F'?

**Review by HanielB — APPROVED** [review]

File: src/expr/proof_rule.h (hunk 5 above)
Comment (ajreynol) [reply src/expr/proof_rule.h:138]: Yes

**Pull request merged**